### PR TITLE
Fuzzy dates

### PR DESCRIPTION
**Pull request by Blackbaud-AlexKingman**
Addresses #74 

Design doc: https://blackbaudinc.app.box.com/s/rxrwahtisziqlc30i6323vwdsp2xxmy8

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -16,7 +16,8 @@ before_install:
   - bash <(curl -s https://blackbaud.github.io/skyux-travis/v3/library-before-install.sh)
 
 install:
-  - bash <(curl -s https://blackbaud.github.io/skyux-travis/v3/library-install.sh)
+  - npm install
```
Comment (Blackbaud-BobbyEarl) [inline .travis.yml]: You should be able to undo this change now Alex.

Comment (Blackbaud-BobbyEarl) [reply .travis.yml]: Sorry, once we're able to merge in the other config changes.

File: src/app/public/modules/datepicker/fuzzy-date.service.ts [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+
+    return fuzzyDate;
+  }
+
+  public getFuzzyDateFromString(date: string, dateFormat: string): SkyFuzzyDate {
+    if (!date || !dateFormat) {
+      return;
+    }
+
+    let day: any;
```
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/fuzzy-date.service.ts]: They can either end up a string or number, based on how the switch/case resolves below. Think that needs addressed to be more strict?

File: src/app/public/modules/datepicker/fuzzy-date.service.ts [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+    if (year && !isNaN(year)) {
+      return parseInt(year, 10);
+    }
+  }
+
+  private isLeapYear(year: number): boolean {
+    return ((year % 4 === 0) && (year % 100 !== 0)) || (year % 400 === 0);
+  }
+
+  private getMonthNumber(month: any): number {
```
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/fuzzy-date.service.ts]: Done.

File: src/app/public/modules/datepicker/fuzzy-date.service.ts [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+      if (!separator && date.indexOf(currentSeparator) > 0) {
+        separator = currentSeparator;
+      }
+    });
+
+    return separator;
+  }
+
+  private get4DigitYearFromDateString(date: string): number {
+    let year: any;
```
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/fuzzy-date.service.ts]: Done.

File: src/app/public/modules/datepicker/fuzzy-date.service.ts [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    }
+
+    return {
+      month: month,
+      day: day,
+      year: year
+    };
+  }
+
+  public getFuzzyDateRange(startFuzzyDate: any, endFuzzyDate: any): SkyFuzzyDateRange {
```
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/fuzzy-date.service.ts]: Yep, done.

File: src/app/public/modules/datepicker/datepicker-input-fuzzy.directive.ts [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+    }
+
+    if (!this.control.value) {
+      return;
+    }
+
+    const value: any = control.value;
+
+    let fuzzyDate: SkyFuzzyDate;
+    let validationError: any;
```
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/datepicker-input-fuzzy.directive.ts]: Done.

File: src/app/public/modules/datepicker/fuzzy-date.service.ts [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    }
+
+    const year = fuzzyDate.year || this.getDefaultYear(fuzzyDate);
+    const month = fuzzyDate.month - 1 || 0;
+    const day = fuzzyDate.day || 1;
+
+    return moment([year, month, day]);
+  }
+
+  public getStringFromFuzzyDate(fuzzyDate: any, dateFormat: string): string {
```
[the remark this answers is not in the document]
Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/datepicker/fuzzy-date.service.ts]: Yep! Done.
